AMDGPU/GlobalISel: Add skeletons for new register bank select passes

petar-avramovic opened the pull request:
New register bank select for AMDGPU will be split in two passes:
- AMDGPURegBankSelect: select banks based on machine uniformity analysis
- AMDGPURegBankLegalize: lower instructions that can't be inst-selected
  with register banks assigned by AMDGPURegBankSelect.
AMDGPURegBankLegalize is similar to legalizer but with context of
uniformity analysis. Does not change already assigned banks.
Main goal of AMDGPURegBankLegalize is to provide high level table-like
overview of how to lower generic instructions based on available target
features and uniformity info (uniform vs divergent).
See RegBankLegalizeRules.

Summary of new features:
At the moment register bank select assigns register bank to output
register using simple algorithm:
- one of the inputs is vgpr output is vgpr
- all inputs are sgpr output is sgpr.
When function does not contain divergent control flow propagating
register banks like this works. In general, first point is still correct
but second is not when function contains divergent control flow.
Examples:
- Phi with uniform inputs that go through divergent branch
- Instruction with temporal divergent use.
To fix this AMDGPURegBankSelect will use machine uniformity analysis
to assign vgpr to each divergent and sgpr to each uniform instruction.
But some instructions are only available on VALU (for example floating
point instructions before gfx1150) and we need to assign vgpr to them.
Since we are no longer propagating register banks we need to ensure that
uniform instructions get their inputs in sgpr in some way.
In AMDGPURegBankLegalize uniform instructions that are only available on
VALU will be reassigned to vgpr on all operands and read-any-lane vgpr
output to original sgpr output.

petar-avramovic commented:
* **#112864** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/112864?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#112863** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/112863?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#112862** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/112862?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @petar-avramovic and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

rovka commented on llvm/lib/Target/AMDGPU/AMDGPU.h:
```diff
@@ -39,6 +39,8 @@ FunctionPass *createSIFoldOperandsLegacyPass();
 FunctionPass *createSIPeepholeSDWALegacyPass();
 FunctionPass *createSILowerI1CopiesLegacyPass();
 FunctionPass *createAMDGPUGlobalISelDivergenceLoweringPass();
+FunctionPass *createAMDGPURBSelectPass();
```
Please spell out RegBank wherever possible, it makes it easier to grep for things in the codebase.

petar-avramovic replied:
Sure, will wait for more comments then bulk update everything

petar-avramovic replied:
Do you have any suggestions since AMDGPURegBankSelect is taken? 
AMDGPURBSelect -> ... 
AMDGPURBLegalize -> AMDGPURegBankLegalize is fine

rovka replied:
Idk, I was hoping we could get rid of the other one :) Maybe we could call this one AMDGPUStandaloneRegBankSelect, since it's not inheriting from RegBankSelect (at least if that's the intention for the long term).

petar-avramovic replied:
The AMDGPURegBankSelect is functional except for some functions with divergent control flow. Deleting it now breaks almost all globalisel regression tests.
When new regbankselect path gets functional we can make it a default path and decide what to do with AMDGPURegBankSelect, maybe rename it to legacy or just delete it.

New pass names as suggested:
 AMDGPUStandaloneRegBankSelect
 AMDGPURegBankLegalize

arsenm replied:
You can delete the old one and replace it with the generic regbankselect pass. It's not doing anything in the pass other than running the analysis

petar-avramovic replied:
Reverting default reg-bank-select to RegBankSelect, new pass will take
AMDGPURegBankSelect

rovka commented on llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:
```diff
@@ -1371,7 +1378,12 @@ void GCNPassConfig::addPreRegBankSelect() {
+    addPass(createAMDGPURBSelectPass());
+    addPass(createAMDGPURBLegalizePass());
+  } else {
+    addPass(new AMDGPURegBankSelect());
```
Could we just modify this AMDGPURegBankSelect instead of adding a new one? It was meant to use uniformity analysis too. Maybe you can propagate the flag into the pass if you want to keep the old behaviour around in the meantime.

petar-avramovic replied:
Older versions were doing that. New passes are much simpler and don't use anything from generic reg-bank-select(except register banks that come from td file via RegisterBankInfo).
To simplify review process and transition period I would prefer two passes separate from AMDGPURegBankSelect, rb-legalize will also be split across few files.
Old behavior is there to keep upcoming patches small by gradually switching to -new-reg-bank-select.

rovka commented on llvm/lib/Target/AMDGPU/AMDGPURBLegalize.cpp:
```diff
@@ -0,0 +1,74 @@
+/// Lower G_ instructions that can't be inst-selected with register bank
+/// assignment given by RB-select based on machine uniformity info.
+/// Given types on all operands, some register bank assignments require lowering
+/// while other do not.
```
```suggestion
/// while others do not.
```

rovka commented on llvm/lib/Target/AMDGPU/AMDGPURBLegalize.cpp:
```diff
@@ -0,0 +1,74 @@
+//===----------------------------------------------------------------------===//
+//
+/// Lower G_ instructions that can't be inst-selected with register bank
+/// assignment given by RB-select based on machine uniformity info.
```
Nit: Maybe mention explicitly that this is expected to run after RB-select? This might be surprising since the usual GISel pipeline has legalization first and then RB-select.

rovka commented on llvm/test/CodeGen/AMDGPU/GlobalISel/regbankselect-mui-rb-legalize.mir:
```diff
@@ -0,0 +1,858 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
+# RUN: llc -mtriple=amdgcn -mcpu=gfx1010 -run-pass=none %s -verify-machineinstrs -o - | FileCheck %s
```
Shouldn't this run rb-legalize? It only returns true, so it should be pretty harmless.

petar-avramovic replied:
true, will update mir test run lines

petar-avramovic replied:
Sorry, correction, not true. Machine verifier will complain that Generic virtual register must have a bank in a RegBankSelected function.
Now that I think of it, RBLegalize should probably also require that MachineFunction has Property::RegBankSelected

arsenm replied:
I'm not sure what the point of this test is then. If it's testing the verifier, it should go in test/MachineVerifier (and then doesn't need -verify-machineinstrs)

petar-avramovic committed:
AMDGPU/GlobalISel: Add skeletons for new register bank select passes

New register bank select for AMDGPU will be split in two passes:
- AMDGPURegBankSelect: select banks based on machine uniformity analysis
- AMDGPURegBankLegalize: lower instructions that can't be inst-selected
  with register banks assigned by AMDGPURegBankSelect.
AMDGPURegBankLegalize is similar to legalizer but with context of
uniformity analysis. Does not change already assigned banks.
Main goal of AMDGPURegBankLegalize is to provide high level table-like
overview of how to lower generic instructions based on available target
features and uniformity info (uniform vs divergent).
See RegBankLegalizeRules.

Summary of new features:
At the moment register bank select assigns register bank to output
register using simple algorithm:
- one of the inputs is vgpr output is vgpr
- all inputs are sgpr output is sgpr.
When function does not contain divergent control flow propagating
register banks like this works. In general, first point is still correct
but second is not when function contains divergent control flow.
Examples:
- Phi with uniform inputs that go through divergent branch
- Instruction with temporal divergent use.
To fix this AMDGPURegBankSelect will use machine uniformity analysis
to assign vgpr to each divergent and sgpr to each uniform instruction.
But some instructions are only available on VALU (for example floating
point instructions before gfx1150) and we need to assign vgpr to them.
Since we are no longer propagating register banks we need to ensure that
uniform instructions get their inputs in sgpr in some way.
In AMDGPURegBankLegalize uniform instructions that are only available on
VALU will be reassigned to vgpr on all operands and read-any-lane vgpr
output to original sgpr output.

arsenm commented on llvm/test/CodeGen/AMDGPU/GlobalISel/regbankselect-mui-salu-float.mir:
```diff
@@ -0,0 +1,92 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
+# RUN: llc -mtriple=amdgcn -mcpu=gfx1010 -run-pass=regbankselect %s -verify-machineinstrs -o - | FileCheck %s -check-prefixes=OLD_RBS_GFX10
+# RUN: llc -mtriple=amdgcn -mcpu=gfx1200 -run-pass=regbankselect %s -verify-machineinstrs -o - | FileCheck %s -check-prefixes=OLD_RBS_GFX12
+# RUN: llc -mtriple=amdgcn -mcpu=gfx1010 -run-pass=regbankselect %s -verify-machineinstrs -o - | FileCheck %s -check-prefixes=NEW_RBS_GFX10
+# RUN: llc -mtriple=amdgcn -mcpu=gfx1200 -run-pass=regbankselect %s -verify-machineinstrs -o - | FileCheck %s -check-prefixes=NEW_RBS_GFX12
```
Don't need -verify-machineinstrs

arsenm replied:
```suggestion
# RUN: llc -mtriple=amdgcn -mcpu=gfx1010 -run-pass=regbankselect %s -o - | FileCheck %s -check-prefixes=OLD_RBS_GFX10
# RUN: llc -mtriple=amdgcn -mcpu=gfx1200 -run-pass=regbankselect %s -o - | FileCheck %s -check-prefixes=OLD_RBS_GFX12
# RUN: llc -mtriple=amdgcn -mcpu=gfx1010 -run-pass=regbankselect %s -o - | FileCheck %s -check-prefixes=NEW_RBS_GFX10
# RUN: llc -mtriple=amdgcn -mcpu=gfx1200 -run-pass=regbankselect %s -o - | FileCheck %s -check-prefixes=NEW_RBS_GFX12
```

arsenm committed:
Remove -verify-machineinstrs
